invoice.go

```diff
@@ -0,0 +1,91 @@
+package ksef
+
+import (
+	"github.com/invopop/gobl/bill"
+	"github.com/invopop/gobl/regimes/common"
+)
+
+type Inv struct {
+	CurrencyCode                       string       `xml:"KodWaluty"`
+	IssueDate                          string       `xml:"P_1"`
+	SequentialNumber                   string       `xml:"P_2"`
+	CompletionDate                     string       `xml:"P_6,omitempty"`
+	StartDate                          string       `xml:"P_6_Od,omitempty"`
+	EndDate                            string       `xml:"P_6_Do,omitempty"`
+	BasicRateNetSale                   string       `xml:"P_13_1,omitempty"`
+	BasicRateTax                       string       `xml:"P_14_1,omitempty"`
+	BasicRateTaxConvertedToPln         string       `xml:"P_14_1W,omitempty"`
+	FirstReducedRateNetSale            string       `xml:"P_13_2,omitempty"`
+	FirstReducedRateTax                string       `xml:"P_14_2,omitempty"`
+	FirstReducedRateTaxConvertedToPln  string       `xml:"P_14_2W,omitempty"`
+	SecondReducedRateNetSale           string       `xml:"P_13_3,omitempty"`
+	SecondReducedRateTax               string       `xml:"P_14_3,omitempty"`
+	SecondReducedRateTaxConvertedToPln string       `xml:"P_14_3W,omitempty"`
+	TotalAmountRecivable               string       `xml:"P_15"`
+	Annotations                        *Annotations `xml:"Adnotacje"`
+	InvoiceType                        string       `xml:"RodzajFaktury"`
+	Lines                              []*Line      `xml:"FaWiersz"`
+}
+
+type Annotations struct {
+	P_16      int `xml:"P_16"`
+	P_17      int `xml:"P_17"`
+	P_18      int `xml:"P_18"`
+	P_18A     int `xml:"P_18A"`
+	P_19N     int `xml:"Zwolnienie>P_19N"`
+	P_22N     int `xml:"NoweSrodkiTransportu>P_22N"`
+	P_23      int `xml:"P_23"`
+	P_PMarzyN int `xml:"PMarzy>P_PMarzyN"`
+}
+
+func NewAnnotations(inv *bill.Invoice) *Annotations {
+	Annotations := &Annotations{ // default values for the most common case
+		P_16:      2,
+		P_17:      2,
+		P_18:      2,
+		P_18A:     2,
+		P_19N:     1,
+		P_22N:     1,
+		P_23:      2,
+		P_PMarzyN: 1,
+	}
+	return Annotations
+}
+
+func NewInv(inv *bill.Invoice) *Inv {
+	cu := inv.Currency.Def().Units
+	Inv := &Inv{
+		Annotations:          NewAnnotations(inv),
+		CurrencyCode:         string(inv.Currency),
+		InvoiceType:          "VAT", // TODO
+		IssueDate:            inv.IssueDate.String(),
+		SequentialNumber:     inv.Series + inv.Code,
+		TotalAmountRecivable: inv.Totals.Payable.Rescale(cu).String(),
+		Lines:                NewLines(inv.Lines),
+	}
+	if inv.OperationDate != nil {
+		Inv.CompletionDate = inv.OperationDate.String()
+	}
+	for _, cat := range inv.Totals.Taxes.Categories {
+		if cat.Code != common.TaxCategoryVAT {
+			continue
+		}
+
+		for _, rate := range cat.Rates {
+			if rate.Key == common.TaxRateStandard {
```

Comment:
This is a tough one... so the `rate.Key` fields in GOBL are actually optional. The user can use a key like `"standard"` to automatically determine the tax rate is 23%, but they could just as easily write "23%" in the `"percent"` property and just ignore the `"key"` field.

Would it make sense to us the % rates instead of keys? You could use a list of comparisons like:
 * If > 15%: standard
 * else if > 6%: Reduced
 * else if > 2%: Super Reduced
 
There is obviously a risk of a significant tax change there breaking the rules, but I think the chances of that are low.

Reply:
I made this change. 
It made the code look somewhat ugly because there is no direct way to compare % values.
Super reduced rate is applied when greater than 4% because there is a special rate for taxi drivers (3 or 4 %, not sure exactly)

Reply:
Is see that in the code! I'll make a suggestion to simplify